### PR TITLE
Refactor RPC to make it more extensible

### PR DESCRIPTION
**Pull request by DanielSchiavini**
### What I did
Changed the NetworkEnv and CachingRPC so they are not dependent on the RPC URL.
This makes it possible to implement a BrowserRPC in #124.

### How I did it
- Added new utility function `boa.set_network_env(url)`
- Changed the `environment` so it has two ways of forking: via an URL or via the RPC
- Changed CachingRPC to support either EthereumRPC or BrowserRPC via an unique identifier
- Deprecated NetworkEnv(url) constructor - Using it with an URL generates a warning

### How to verify it
- `boa.set_network_env(url)` should work as previously
- `boa.env.fork(url)` should work as previously
- `boa.env_fork_rpc(EthereumRPC(url))` should do the same as `boa.env.fork(url)`
- `NetworkEnv(url)` should work as previously, however it will generate a deprecation warning

### Description for the changelog
- Deprecated NetworkEnv(url) constructor - Using it with an URL generates a warning.

### Cute Animal Picture

image

### CODE REVIEW

File: boa/vm/fork.py [hunk 1]
```diff
@@ -115,19 +126,12 @@ def __init__(self, *args, **kwargs):
...
-            "earliest",
-        ):
+        self._rpc = CachingRPC(self._rpc, **rpc_kwargs)
+
+        if block_identifier not in _PREDEFINED_BLOCKS:
             block_identifier = to_hex(block_identifier)
 
-        # do not cache - use raw_fetch
-        self._block_info = self._rpc._raw_fetch_single(
+        self._block_info = self._rpc.fetch(
```
Comment (charles-cooper) [inline boa/vm/fork.py:134]: hm i'm not sure we can make this change -- doesn't CachingRPC cache the result?

Comment (DanielSchiavini) [reply boa/vm/fork.py]: I've added a fetch_uncached method to address your concern

File: boa/vm/fork.py [hunk 2]
```diff
@@ -90,6 +90,10 @@ def fetch(self, method, params):
         (res,) = self.fetch_multi([(method, params)])
         return res
 
+    def fetch_uncached(self, method, params):
+        # due to __new__, underlying rpc is never CachingRPC
+        return self._rpc.fetch(method, params)
```
Comment (charles-cooper) [inline boa/vm/fork.py:95]: we could make it part of the `RPC` api, and then i think here we just dispatch to `self._rpc.fetch_uncached`

Comment (DanielSchiavini) [reply boa/vm/fork.py:95]: This is only applicable to the CachingRPC, the other ones do not cache anything.
Also note the CachingRPC is only used in the AccountFork and it's always guaranteed to be a CachingRPC.
Why do you think this should be part of the RPC API?

Comment (charles-cooper) [reply boa/vm/fork.py:95]: > Why do you think this should be part of the RPC API?
it makes the implementation slightly cleaner (we don't need that comment anymore!). i also think it's pretty clear what it's for, like any RPC implementation might be free to cache results, and if you really want the uncached result you can use `fetch_uncached`.

Comment (DanielSchiavini) [reply boa/vm/fork.py]: I've lifted the implementation to the base class, but the comment is still needed atm

File: boa/environment.py [hunk 3]
```diff
@@ -439,6 +441,8 @@ def _init_vm(self, reset_traces=True):
         self.vm = self.chain.get_vm()
 
         self.vm.patch = VMPatcher(self.vm)
+        # revert any previous AccountDBFork patching
+        self.vm.__class__._state_class.account_db_class = AccountDB
```
Comment (charles-cooper) [inline boa/environment.py:445]: i'd prefer to move this to a property of the environment

**Review by charles-cooper — APPROVED** [review]
lgtm, thanks!

**Pull request merged**